Add basic array methods

davidot opened the pull request:
Most of these were easy pickings
But I did run into a couple of interesting cases so I have another couple of questions

davidot commented on include/natalie/array_value.hpp:
```diff
@@ -110,7 +113,8 @@ class ArrayValue : public Value {
     ValuePtr push(Env *, size_t, ValuePtr *);
-    ValuePtr ref(Env *, ValuePtr, ValuePtr);
+    ValuePtr rassoc(Env *, ValuePtr);
+    ValuePtr ref(Env *, ValuePtr, ValuePtr = nullptr);
```
Giving the second argument a default value make it easy to use as a binding
But is that what we want? Or would a method `at` which just calls `ref` be better?

seven1m replied:
I think I prefer the latter (method `at` which calls `ref`), just for clarity of purpose. :smile:

davidot replied:
Made it an explicit method

davidot commented on src/array_value.cpp:
```diff
@@ -257,6 +257,25 @@ ValuePtr ArrayValue::first(Env *env, ValuePtr n) {
 }
 
+
+ValuePtr ArrayValue::drop(Env *env, ValuePtr n) {
```
A lot of duplicated logic but let first implement everything before we refactor and optimize

davidot commented on src/array_value.cpp:
```diff
@@ -507,4 +526,43 @@ ValuePtr ArrayValue::uniq(Env *env) {
 }
 
+ValuePtr ArrayValue::clear(Env *env) {
+    this->assert_not_frozen(env);
```
Noticed `assert_not_frozen` in pop when should this be called?
Anytime when changing the underlying value?

seven1m replied:
Yes, any time we are mutating the underlying object rather than returning a copy, we should add `assert_not_frozen`.

davidot commented on test/natalie/array_test.rb:
```diff
@@ -475,6 +475,15 @@
     end
   end
 
+  describe '#collect' do
+      it 'returns a new array of the result of evaluating the block for each item in the array' do
+        result = [1, 2, 3].collect do |i|
+          i * 2
+        end
+        result.should == [2, 4, 6]
+      end
+    end
+
```
This is just a copy and paste of the `#map` test is there some easier way to test aliases?
Since now you would have to double test writing as well (or you might just forget something has an alias

seven1m replied:
I supposed you could do something like this:

```ruby
['map', 'collect'].each do |method_name|
  describe "##{method_name}" do
    it 'whatever' do
      result = [1, 2, 3].send(method_name) do |i|
        i * 2
      end
      result.should == [2, 4, 6]
    end
  end
end
```

BUT I'm not sure that's better. I think I'd rather just have the duplication. :man_shrugging:

davidot replied:
Alright, kept it like this

davidot commented on test/natalie/array_test.rb:
```diff
@@ -658,6 +700,96 @@
+  describe '#clear' do
+    it 'should leave an empty array empty' do
+      a = []
+      (a.clear).should === a
```
Is this how you test the value of an expression?
And do we want to actually test / specify values of expressions?

seven1m replied:
I don't think this would test what you want. The left side would be evaluated before the right side is passed into `should ===` Also, `===` in Ruby is kinda special for use by `case/when`, so probably not appropriate here.

I think you just want:

```
a = []
a.clear.should == []
a.should == []
```

seven1m replied:
[Note: I deleted an earlier comment here that wasn't right.]

This is close, but `===` isn't what you want here. You should use `a.clear.should equal(a)` since that tests for object identity (same pointer).

If `===` is doing that currently, then it's a bug :wink:

davidot replied:
Changed it to `equal(a)` but this did work in both ruby and natalie so not sure what that means?

seven1m replied:
Ahh, you're right. I'm so used to seeing `===` used for case/when that it looked wrong to me. Carry on :smile:

davidot commented:
I think I addressed all the comments.
Maybe another question, should array_test.rb be split up at some point? It is already quite large but then testing does become more difficult of course

seven1m commented:
Yeah, if you think we should, we can split out tests for each method into their own files inside an `array` directory:

```
array/
  at_test.rb
  assoc_test.rb
  clear_test.rb
  ...
```

I'm super happy with this work, so I'm going to go ahead and merge. <3
